bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/api/utils/TypeUtils.java

```diff
@@ -151,11 +152,28 @@ public static boolean isSameType(Type sourceType, Type targetType) {
      * @return the referred type if provided with a type reference type, else returns the original type
      */
     public static Type getReferredType(Type type) {
-        Type constraint = type;
         if (type.getTag() == TypeTags.TYPE_REFERENCED_TYPE_TAG) {
-            constraint = getReferredType(((ReferenceType) type).getReferredType());
+            return getReferredType(((ReferenceType) type).getReferredType());
         }
-        return constraint;
+        return type;
     }
 
+    /**
+     * Retrieve the conclusive type if a given type is a type reference type or an intersection.
+     *
+     * @param type type to retrieve referred
+     * @return the conclusive type if provided with a type reference type or an intersection,
+     * else returns the original type
+     */
+    public static Type getConclusiveType(Type type) {
```

Comment:
Do we need to expose this method? Can we add this to internal package?

Reply:
Do we need to dereprecate getReferredType method?

Reply:
`getConclusiveType` is useful for the consumer to handle intersection types. Otherwise, they have to manually get the effective type. We are not going to deprecate `getReferredType` since the purpose of `getReferredType` and `getConclusiveType` is different